File: ctui/src/gllib/glmath.c
const float CGUI_ONE_DEG_IN_RAD = 0.0174533f;
float cgui_deg_to_rad(float degrees) { return degrees * 0.0174533f; }


void cgui_mul_mat4(struct Mat4* mat, struct Mat4* a, struct Mat4* b)
{	for (size_t i = 0; i < 4; i++)
	{	for (size_t j = 0; j < 4; j++)
		{	mat->data[4 * j + i] = 0.0f;

			for (size_t k = 0; k < 4; k++)
			{	mat->data[4 * j + i] += a->data[4 * k + i] * b->data[4 * j + k]; } } } }

// fovy: Vertical field of view angle in degrees.
// aspect: Aspect ratio of the viewport. It's calculated as width / height.
// z_near: Objects closer than this value will not be visible.
// z_far: Objects farther away than this value will not be visible.
void cgui_perspective
(	struct Mat4* mat,
    float fovy, float aspect,
    float z_near, float z_far)
{	float tan_half_fovy = tan(fovy / 2.0f);

	cgui_zero_mat4(mat);
	mat->data[0] = 1.0f / (aspect * tan_half_fovy);
	mat->data[4 * 1 + 1] = 1.0f / tan_half_fovy;
	mat->data[4 * 2 + 2] = -(z_far + z_near) / (z_far - z_near);
	mat->data[4 * 2 + 3] = -1.0f;
	mat->data[4 * 3 + 2] = -(2.0f * z_far * z_near) / (z_far - z_near); }


void cgui_add_vec3f(struct Vec3f* vec, struct Vec3f* a, struct Vec3f* b)
{	vec->x = a->x + b->x;
	vec->y = a->y + b->y;
	vec->z = a->z + b->z; }


void cgui_sub_vec3f(struct Vec3f* vec, struct Vec3f* a, struct Vec3f* b)
{	vec->x = a->x - b->x;
	vec->y = a->y - b->y;
	vec->z = a->z - b->z; }


void cgui_normalize_vec3f(struct Vec3f* vec)
{	if (0.0f == vec->x + vec->y + vec->z)
	{	printf("cgui_normalize_vec3f: Zeroed vector found! Skipping...");
		return; }

	float length =
	    sqrt(vec->x * vec->x
	        + vec->y * vec->y
	        + vec->z * vec->z);
	float inv_length = 1.0f / length;
	vec->x *= inv_length;
	vec->y *= inv_length;
	vec->z *= inv_length; }


void cgui_cross(struct Vec3f* vec, struct Vec3f* a, struct Vec3f* b)
{	vec->x = a->y * b->z
	    - a->z * b->y;
	vec->y = a->z * b->x
	    - a->x * b->z;
	vec->z = a->x * b->y
	    - a->y * b->x; }


float cgui_dot(struct Vec3f* a, struct Vec3f* b)
{	return a->x * b->x
	    + a->y * b->y
	    + a->z * b->z; }


void cgui_lookat
(	struct Mat4* mat,
    struct Vec3f* pos_camera,
    struct Vec3f* target,
    struct Vec3f* i_up)
{	struct Vec3f forward;
	cgui_sub_vec3f(&forward, target, pos_camera);
	cgui_normalize_vec3f(&forward);

	struct Vec3f right;
	cgui_cross(&right, &forward, i_up);
	cgui_normalize_vec3f(&right);

	struct Vec3f up;
	cgui_cross(&up, &right, &forward);

	cgui_set_diagonal_mat4(mat, 1.0f);
	mat->data[0] = right.x;
	mat->data[4 * 1] = right.y;
	mat->data[4 * 2] = right.z;
	mat->data[1] = up.x;
	mat->data[4 * 1 + 1] = up.y;
	mat->data[4 * 2 + 1] = up.z;
	mat->data[2] = -forward.x;
	mat->data[4 * 1 + 2] = -forward.y;
	mat->data[4 * 2 + 2] = -forward.z;
	mat->data[4 * 3] = -cgui_dot(&right, pos_camera);
	mat->data[4 * 3 + 1] = -cgui_dot(&up, pos_camera);
	mat->data[4 * 3 + 2] = cgui_dot(&forward, pos_camera); }


// Let's assune that we want 1 meter canva from 16:9 monitor.
//	struct Vec3f camera_position;
//	struct Vec3f camera_target;
//	struct Vec3f camera_up;
//	cgui_set_vec3f_from_floats(&camera_position, 0.0f, -0.0f, -1.0f);
//	cgui_set_vec3f_from_floats(&camera_target, 0.0f, 0.0f, 0.0f);
//	cgui_set_vec3f_from_floats(&camera_up, 0.0f, 1.0f, 0.0f);
//
//	struct Mat4 view2;
//	cgui_lookat(&view2, &camera_position, &camera_target, &camera_up);
//	float left = -0.5f; // Half a meter from middle to the left
//	float right = 0.5f; // Half a meter from middle to the right
//	float bottom = -0.28125f; // 16:9 aspect ratio, so -0.5 * 9 / 16 = -0.28125
//	float top = 0.28125f; // 16:9 aspect ratio, so 0.5 * 9 / 16 = 0.28125
//	float near = 0.01f; // Anything closer than this will clip
//	float far = 4000.0f; // Anything further than this will clip
//
//	struct Mat4 ortho_proj;
//	cgui_ortho(&ortho_proj, left, right, bottom, top, near, far);
//
//	struct Mat4 ortho_view_proj;
//	cgui_mul_mat4(&ortho_view_proj, &ortho_proj, &view2);
void cgui_ortho
(	struct Mat4* mat,
    float left,
    float right,
    float bottom,
    float top,
    float z_near,
    float z_far)
{	cgui_zero_mat4(mat);
	cgui_set_diagonal_mat4(mat, 1.0f);
	mat->data[0] = 2.0f / (right - left);
	mat->data[4 * 1 + 1] = 2.0f / (top - bottom);
	mat->data[4 * 2 + 2] = -2.0f / (z_far - z_near);
	mat->data[4 * 3] = -(right + left) / (right - left);
	mat->data[4 * 3 + 1] = -(top + bottom) / (top - bottom);
	mat->data[4 * 3 + 2] = -(z_far + z_near) / (z_far - z_near); }

File: GLMLib/glmath.f90
module GLModule
	implicit none
	type :: Vec2i
		integer :: x, y
	end type Vec2i
	! There are no unsigned in Fortran
	type :: Vec2f
		real :: x, y
	end type Vec2f
	type :: Vec2d
		double precision :: x, y
	end type Vec2d

	type :: Vec3i
		integer :: x, y, z
	end type Vec3i
	type :: Vec3f
		real :: x, y, z
	end type Vec3f
	type :: Vec3f64
		double precision :: x, y, z
	end type Vec3f64

	type :: Col8
		integer(1) :: r, g, b
	end type Col8
	type :: Col8a
		integer(1) :: r, g, b, a
	end type Col8a
	type :: Col5
		integer(2) :: rgba
	end type Col5

	! The reason for 1D is easier to expend with multiplications
	! 2D flattening will require divisions and modulo,
	! which is slower than multiplications
	type :: Mat4
		real :: data(16)
	end type Mat4
end module GLModule


subroutine set_vec3f_from_floats(vec, x, y, z)
	use GLModule
	type(Vec3f), intent(out) :: vec
	real, intent(in) :: x, y, z

	vec%x = x
	vec%y = y
	vec%z = z
end subroutine set_vec3f_from_floats


subroutine set_vec3f_from_vec3f(vec, src)
	use GLModule
	type(Vec3f), intent(out) :: vec
	type(Vec3f), intent(in) :: src

	vec%x = src%x
	vec%y = src%y
	vec%z = src%z
	end subroutine set_vec3f_from_vec3f


subroutine mul_mat4(mat, a, b)
	use GLModule
	type(Mat4), intent(out) :: mat
	type(Mat4), intent(in) :: a, b
	integer :: i, j, k

	! Not a good idea to do loop fusion on this
	do i = 1, 16
		mat%data(i) = 0.0
	end do

	do i = 1, 4
		do j = 1, 4
			do k = 1, 4
				mat%data(4 * j + i) = mat%data(4 * j + i) + (a%data(4 * k + i) * b%data(4 * j + k))
			end do
		end do
	end do
	end subroutine mul_mat4


! fovy: Vertical field of view angle in degrees.
! aspect: Aspect ratio of the viewport. It's calculated as width / height.
! z_near: Objects closer than this value will not be visible.
! z_far: Objects farther away than this value will not be visible.
subroutine perspective(mat, fovy, aspect, z_near, z_far)
	use GLModule
	type(Mat4), intent(out) :: mat
	real, intent(in) :: fovy, aspect, z_near, z_far
	integer :: i
	real :: tan_half_fovy

	tan_half_fovy = tan(fovy / 2.0);

	do i = 1, 16
		mat%data(i) = 0.0
	end do

	mat%data(1) = 1.0 / (aspect * tan_half_fovy)
	mat%data(4 * 1 + 2) = 1.0 / tan_half_fovy
	mat%data(4 * 2 + 3) = -(z_far + z_near) / (z_far - z_near)
	mat%data(4 * 2 + 4) = -1.0
	mat%data(4 * 3 + 3) = -(2.0 * z_far * z_near) / (z_far - z_near)
end subroutine perspective


subroutine add_vec3f(vec, a, b)
	use GLModule
	type(Vec3f), intent(out) :: vec
	type(Vec3f), intent(in) :: a, b

	vec%x = a%x + b%x
	vec%y = a%y + b%y
	vec%z = a%z + b%z
end subroutine add_vec3f


subroutine sub_vec3f(vec, a, b)
	use GLModule
	type(Vec3f), intent(out) :: vec
	type(Vec3f), intent(in) :: a, b

	vec%x = a%x - b%x
	vec%y = a%y - b%y
	vec%z = a%z - b%z
end subroutine sub_vec3f


subroutine normalize_vec3f(vec)
	use GLModule
	type(Vec3f), intent(out) :: vec
	real :: length, inv_length

	if (abs(vec%x + vec%y + vec%z) < 1.0e-6) then
		write(*, *) "normalize_vec3f: Zeroed vector found! Skipping..."
		return
	endif

	length = &
		sqrt(vec%x * vec%x &
			+ vec%y * vec%y &
			+ vec%z * vec%z)
	inv_length = 1.0 / length
	vec%x = vec%x * inv_length
	vec%y = vec%y * inv_length
	vec%z = vec%z * inv_length
end subroutine normalize_vec3f


subroutine cross(vec, a, b)
	use GLModule
	type(Vec3f), intent(out) :: vec
	type(Vec3f), intent(in) :: a, b

	vec%x = a%y * b%z - a%z * b%y
	vec%y = a%z * b%x - a%x * b%z
	vec%z = a%x * b%y - a%y * b%x
end subroutine cross


subroutine dot(ret, a, b)
	use GLModule
	real, intent(out) :: ret
	type(Vec3f), intent(in) :: a, b

	ret = a%x * b%x &
		+ a%y * b%y &
		+ a%z * b%z
end subroutine dot


subroutine lookat(mat, pos_camera, look_at, i_up)
	use GLModule
	type(Mat4), intent(out) :: mat
	type(Vec3f), intent(in) :: pos_camera, look_at, i_up
	type(Vec3f) :: forward, right, up
	interface
		subroutine normalize_vec3f(vec)
			use GLModule
			type(Vec3f), intent(out) :: vec
		end subroutine normalize_vec3f

		subroutine cross(vec, a, b)
			use GLModule
			type(Vec3f), intent(out) :: vec
			type(Vec3f), intent(in) :: a, b
		end subroutine cross

		subroutine dot(ret, a, b)
			use GLModule
			real, intent(out) :: ret
			type(Vec3f), intent(in) :: a, b
		end subroutine dot
	end interface


	forward%x = look_at%x - pos_camera%x
	forward%y = look_at%y - pos_camera%y
	forward%z = look_at%z - pos_camera%z
	call normalize_vec3f(forward)

	call cross(right, forward, i_up)
	call normalize_vec3f(right)

	call cross(up, right, forward)

	mat%data(1) = right%x
	mat%data(2) = up%x
	mat%data(3) = -forward%x
	mat%data(4) = 0.0

	mat%data(4 * 1 + 1) = right%y
	mat%data(4 * 1 + 2) = up%y
	mat%data(4 * 1 + 3) = -forward%y
	mat%data(4 * 1 + 4) = 0.0

	mat%data(4 * 2 + 1) = right%z
	mat%data(4 * 2 + 2) = up%z
	mat%data(4 * 2 + 3) = -forward%z
	mat%data(4 * 2 + 4) = 0.0

	call dot(mat%data(4 * 3 + 1), right, pos_camera)
	call dot(mat%data(4 * 3 + 2), up, pos_camera)
	mat%data(4 * 3 + 1) = -mat%data(4 * 3 + 1)
	mat%data(4 * 3 + 2) = -mat%data(4 * 3 + 2)
	call dot(mat%data(4 * 3 + 3), forward, pos_camera)
	mat%data(4 * 3 + 4) = 0.0
end subroutine lookat


subroutine ortho(mat, left, right, bottom, top, z_near, z_far)
	use GLModule
	type(Mat4), intent(out) :: mat
	real, intent(in) :: left, right, bottom, top, z_near, z_far

	mat%data(1) = 2.0 / (right - left)
	mat%data(2) = 0.0
	mat%data(3) = 0.0
	mat%data(4) = 0.0

	mat%data(4 * 1 + 1) = 0.0
	mat%data(4 * 1 + 2) = 2.0 / (top - bottom)
	mat%data(4 * 1 + 3) = 0.0
	mat%data(4 * 1 + 4) = 0.0

	mat%data(4 * 2 + 1) = 0.0
	mat%data(4 * 2 + 2) = 0.0
	mat%data(4 * 2 + 3) = -2.0 / (z_far - z_near)
	mat%data(4 * 2 + 4) = 0.0

	mat%data(4 * 3 + 1) = -(right + left) / (right - left)
	mat%data(4 * 3 + 2) = -(top + bottom) / (top - bottom)
	mat%data(4 * 3 + 3) = -(z_far + z_near) / (z_far - z_near)
	mat%data(4 * 3 + 4) = 1.0
end subroutine ortho
